## Task 1 - 1 hour

In [1]:
# import libraries

import numpy as np
import pandas as pd

### 1.1. Download the data set movie_metadata.csv, which contains data about films from IMDb (Internet Movie Database).

In [2]:
# read csv file into dataframe
df = pd.read_csv("./data/movie_metadata.csv")
df.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4    NaN        Doug Walker                     NaN       NaN   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    131.0                     NaN        Rob Walker   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   131.0          NaN                      Documentary  ...   

  num_user_for_reviews language  country  content_rating        budget  \
0               3054.0  English      USA           PG-13  237000000.0$   
1               1238.0  English      USA           PG-13  300000000.0$   
2                994.0  English       UK           PG-13  245000000.0$   
3               2701.0  English      USA           PG-13  250000000.0$   
4                  NaN      NaN      NaN             NaN          0.0$   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4         NaN                   12.0        7.1           NaN   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                    0  

[5 rows x 28 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5043 entries, 0 to 5042
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      5024 non-null   object 
 1   director_name              4939 non-null   object 
 2   num_critic_for_reviews     4993 non-null   float64
 3   duration                   5028 non-null   float64
 4   director_facebook_likes    4939 non-null   float64
 5   actor_3_facebook_likes     5020 non-null   float64
 6   actor_2_name               5030 non-null   object 
 7   actor_1_facebook_likes     5036 non-null   float64
 8   gross                      4159 non-null   float64
 9   genres                     5043 non-null   object 
 10  actor_1_name               5036 non-null   object 
 11  movie_title                5043 non-null   object 
 12  num_voted_users            5043 non-null   int64  
 13  cast_total_facebook_likes  5043 non-null   int64

### 1.2. The duration column contains data on the film length. How many missing values are there in this column?

15 Missing Values

In [4]:
# Finding missing values using .isnull() method
duration_null = df.duration.isnull()
duration_null.value_counts()

False    5028
True       15
Name: duration, dtype: int64

### 1.3. Replace the missing values in the duration column with the median value for this column.

In [5]:
median_duration = df.duration.median()
df.duration.fillna(median_duration, inplace=True)

### 1.4. What is the average film length? Give the answer as a floating-point figure rounded to two decimal places.

In [6]:
average_duration = round(df.duration.mean(),2)
average_duration

107.19

### 1.5. Create a movie_duration_category column, which will contain three categories depending on the film length:
• Category "1. <90" if the film is less than 90 minutes long

• Category "2. 90–120" if the film is between 90 minutes and two hours long (inclusively)

• Category "3. >120" if the film is more than two hours long

In [7]:
# Defining a function to categorise each duration
def new_duration_category(df: pd.DataFrame):
    new_values = []
    for d in df.duration.to_list():
        if d < 90:
            new_values.append("<90")
        elif d >= 90 and d <= 120:
            new_values.append("90-120")
        else:
            new_values.append(">120")
    return new_values

In [8]:
# Creating new category
df["duration_category"] = new_duration_category(df)

### 1.6. Build a summary table for films released after 2000 (inclusively), to list the numbers of films:
• Table rows: year

• Table columns: movie duration category ("<90", "90–120", ">120")

• The year of release should be displayed in the YYYY format.

In [9]:
# Setting title year as a pandas datetime object YYYY
df["title_year_all"] = pd.to_datetime(df["title_year"], format="%Y")
df["title_year"] = df.title_year_all.dt.year

In [12]:
df[df["title_year"] >= 2000].describe(include="all", datetime_is_numeric=True)

color      director_name  num_critic_for_reviews     duration  \
count    3583               3597             3559.000000  3597.000000   
unique      2               1982                     NaN          NaN   
top     Color  Steven Soderbergh                     NaN          NaN   
freq     3492                 13                     NaN          NaN   
mean      NaN                NaN              164.099747   106.326105   
min       NaN                NaN                1.000000     7.000000   
25%       NaN                NaN               65.000000    93.000000   
50%       NaN                NaN              138.000000   103.000000   
75%       NaN                NaN              230.000000   116.000000   
max       NaN                NaN              813.000000   300.000000   
std       NaN                NaN              131.142289    20.084571   

        director_facebook_likes  actor_3_facebook_likes    actor_2_name  \
count               3597.000000             3581.000000            3587   
unique                      NaN                     NaN            2296   
top                         NaN                     NaN  Morgan Freeman   
freq                        NaN                     NaN              13   
mean                 560.721435              724.050824             NaN   
min                    0.000000                0.000000             NaN   
25%                    7.000000              131.000000             NaN   
50%                   46.000000              392.000000             NaN   
75%                  180.000000              655.000000             NaN   
max                23000.000000            23000.000000             NaN   
std                 2528.497778             1890.474136             NaN   

        actor_1_facebook_likes         gross genres  ... country  \
count              3590.000000  3.070000e+03   3597  ...    3596   
unique                     NaN           NaN    733  ...      60   
top                        NaN           NaN  Drama  ...     USA   
freq                       NaN           NaN    170  ...    2660   
mean               7201.938440  4.884831e+07    NaN  ...     NaN   
min                   0.000000  1.620000e+02    NaN  ...     NaN   
25%                 629.250000  4.436326e+06    NaN  ...     NaN   
50%                1000.000000  2.499434e+07    NaN  ...     NaN   
75%               12000.000000  6.123895e+07    NaN  ...     NaN   
max              640000.000000  7.605058e+08    NaN  ...     NaN   
std               16923.476267  7.088449e+07    NaN  ...     NaN   

       content_rating  budget   title_year actor_2_facebook_likes  \
count            3358    3597  3597.000000            3587.000000   
unique             10     349          NaN                    NaN   
top                 R    0.0$          NaN                    NaN   
freq             1468     335          NaN                    NaN   
mean              NaN     NaN  2008.117320            1844.585169   
min               NaN     NaN  2000.000000               0.000000   
25%               NaN     NaN  2004.000000             290.500000   
50%               NaN     NaN  2008.000000             625.000000   
75%               NaN     NaN  2012.000000             953.500000   
max               NaN     NaN  2016.000000          137000.000000   
std               NaN     NaN     4.658066            4432.766853   

         imdb_score aspect_ratio movie_facebook_likes  duration_category  \
count   3597.000000  3320.000000          3597.000000               3597   
unique          NaN          NaN                  NaN                  3   
top             NaN          NaN                  NaN             90-120   
freq            NaN          NaN                  NaN               2384   
mean       6.311732     2.184666          9431.330275                NaN   
min        1.600000     1.330000             0.000000                NaN   
25%        5.700000     1.850000             0.00000

### 1.7. How many films between 90 minutes and two hours long were released in 2008?

In [11]:
df_ge_2000 = df[df["title_year"] >= 2000]

In [16]:
df[df["title_year"] >= 2008]["duration_category"].value_counts()

90-120    1338
>120       354
<90        290
Name: duration_category, dtype: int64

### Answer is 1338 films

### 1.8. The plot_keywords column holds keywords characterizing the film's plot. Using the data in this column, create a column called movie_plot_category, to contain four categories depending on the key words in the column:
• Category "love_and_death" if the keywords include both "love" and "death"

• Category "love" if the keywords include the word "love"

• Category "death" if the keywords include the word "death"

• Category "other" if the keywords do not meet the conditions above

In [13]:
# classify plot_keywords into specified category using a function call
def new_plot_category(df: pd.DataFrame):
    new_values = []
    for d in df.plot_keywords.to_list():
        try:
            if "love" in d and "death" in d:
                new_values.append("love_and_death")
            elif "love" in d:
                new_values.append("love")
            elif "death" in d:
                new_values.append("death")
            else:
                new_values.append("other")
                
        except TypeError:
            new_values.append("other")
            continue
    return new_values

In [14]:
df["movie_plot_category"] = new_plot_category(df)

In [15]:
imdb_rating_avg = df.groupby("movie_plot_category")["imdb_score"].mean()

### 1.9. The imdb_score column shows a viewer rating for the film. Build a table to reflect the average rating of films depending on which movie_plot_category category they belong to.

In [16]:
imdb_rating_avg

movie_plot_category
death             6.535465
love              6.580769
love_and_death    6.506250
other             6.431422
Name: imdb_score, dtype: float64

### 1.10. What is the average rating of films in the "love" category? Give the answer as a floating point figure rounded to two decimal places.

In [34]:
round(imdb_rating_avg.love, 2)

6.58

### answer is 6.58

### 1.11. The budget column contains the film's budget. What is the median budget for all the films listed? Give the answer as an integer.

In [17]:
# slicing data values in budget column to find the median
for i, d in df.budget.items():
    df["budget"][i] = d[:-3]

/tmp/ipykernel_33527/3626028471.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["budget"][i] = d[:-3]


In [18]:
median_budget = int(df.budget.median())
median_budget

15000000

### Answer is $15,000,000

## Task 2 > 3 hours

### 1. Download the event_data.csv dataset, which contains data on the use of the mobile application of users who registered from July 29 to September 1, 2019:
• user_id - user identifier;

• event_date - time of the event;

• event_type - type of event: registration - registration in the application; 
simple_event - click event in the application; purchase - an event of purchase within the application; purchase_amount - purchase amount.

In [20]:
# Read csv file
dfe = pd.read_csv("./data/event_data.csv")
dfe.head()

user_id           event_date    event_type  purchase_amount
0  c40e6a  2019-07-29 00:02:15  registration              NaN
1  a2b682  2019-07-29 00:04:46  registration              NaN
2  9ac888  2019-07-29 00:13:22  registration              NaN
3  93ff22  2019-07-29 00:16:47  registration              NaN
4  65ef85  2019-07-29 00:19:23  registration              NaN

### 2. Highlight user cohorts based on the week of registration in the application. The cohort identifier should be the week ordinal (for example, the week from July 29 to August 4 should have identifier 31).

In [21]:
# Highlighting the registration week as a cohort identifier
dfe["event_date"] = pd.to_datetime(dfe["event_date"], infer_datetime_format=True)
dfe["week"] = dfe["event_date"].dt.isocalendar().week

In [23]:
dfe.head(20)

user_id          event_date    event_type  purchase_amount  week
0   c40e6a 2019-07-29 00:02:15  registration              NaN    31
1   a2b682 2019-07-29 00:04:46  registration              NaN    31
2   9ac888 2019-07-29 00:13:22  registration              NaN    31
3   93ff22 2019-07-29 00:16:47  registration              NaN    31
4   65ef85 2019-07-29 00:19:23  registration              NaN    31
5   90852e 2019-07-29 00:21:16  registration              NaN    31
6   357151 2019-07-29 00:25:53  registration              NaN    31
7   71ac11 2019-07-29 00:28:51  registration              NaN    31
8   af679d 2019-07-29 00:30:46  registration              NaN    31
9   a48f29 2019-07-29 00:41:54  registration              NaN    31
10  b65930 2019-07-29 00:47:22  registration              NaN    31
11  956ad6 2019-07-29 00:54:13  registration              NaN    31
12  8aa5b4 2019-07-29 00:58:22  registration              NaN    31
13  5fb555 2019-07-29 01:07:24  registration              NaN    31
14  37fa41 2019-07-29 01:10:32  registration              NaN    31
15  b5787e 2019-07-29 01:11:22  registration              NaN    31
16  b2e16e 2019-07-29 01:14:16  registration              NaN    31
17  ca3c58 2019-07-29 01:18:51  registration              NaN    31
18  bea18b 2019-07-29 01:41:59  registration              NaN    31
19  48cac1 2019-07-29 01:42:46  registration              NaN    31

### 3. How many unique users in the cohort with ID 33?

In [62]:
dfe[dfe.week == 33].shape[0]

18144

### Answer is 18144 unique users

#

### 4. For each event, highlight the indicator lifetime - the weekly lifetime of the cohort. The lifetime indicator is calculated based on the serial number of the week in which the event  is committed, relative to the week of registration. For example, an event committed on August 3 by a user from a cohort of registrants at 31 weeks will be committed on the zero week of lifetime, and an event committed by the same user on August 5 will be committed on the first week of lifetime).

In [59]:
# grouping user_id and event type to get unique entries for each
df_lifetime = dfe.groupby(["week","user_id", "event_type"]).count()

In [61]:
df_lifetime.reset_index(inplace=True)

In [62]:
df_lifetime["lifetime"] = 0

In [66]:
# getting the lifetime count for each user per event type
some_list = list()
for i, v in df_lifetime.iterrows():
    temp_list = [v["user_id"], v["event_type"]]
    if temp_list in some_list:
        n = some_list.count(temp_list)
        df_lifetime["lifetime"][i] = n
        some_list.append(temp_list)
    else:
        df_lifetime["lifetime"][i] == 0
        some_list.append(temp_list)

/tmp/ipykernel_33527/2476522679.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_lifetime["lifetime"][i] = n


In [75]:
df_lifetime = df_lifetime[["week", "user_id", "event_type","lifetime"]]

### Result of Operation

In [79]:
df_lifetime[df_lifetime["week"] == 34].head(20)

week user_id    event_type  lifetime
17794    34  00049f  simple_event         0
17795    34  00082c  simple_event         3
17796    34  001d5d  registration         0
17797    34  002195  simple_event         1
17798    34  004304  registration         0
17799    34  004cd2  simple_event         1
17800    34  004dc9  simple_event         1
17801    34  004fcb  simple_event         0
17802    34  00505f  simple_event         3
17803    34  006139      purchase         0
17804    34  006139  registration         0
17805    34  006139  simple_event         0
17806    34  006657  simple_event         0
17807    34  00675c      purchase         0
17808    34  00675c  registration         0
17809    34  00675c  simple_event         0
17810    34  006867  simple_event         0
17811    34  0069d0      purchase         0
17812    34  0069d0  simple_event         1
17813    34  006f31  registration         0

### 5. Build a summary table of changes in the Retention Rate for cohorts depending on lifetime.

In [82]:
retention_count = df_lifetime["week"].value_counts().to_frame()
retention_count

week
35  9278
34  8730
33  7871
32  6166
31  3757

In [102]:
df_retention = df_lifetime[df_lifetime.lifetime > 0].copy()
df_retention = df_retention.groupby("week")["lifetime"].count().to_frame()
df_retention

lifetime
week          
32        1420
33        2812
34        3710
35        4102

In [98]:
b = [2,4,6,1]
b[::-1]

[1, 6, 4, 2]

In [103]:
sum_lifetime = df_retention.lifetime.to_list()
count_lifetime = list(retention_count.week.to_list())
count_lifetime = count_lifetime[::-1]
count_lifetime

[3757, 6166, 7871, 8730, 9278]

In [104]:
count = 0
reten = list()
for n, z in zip(sum_lifetime, count_lifetime[1:]):
    if n == 0:
        reten.append(100.0)
        count += 1
        continue
    percentage = (n/z) * 100
    reten.append(percentage)
    

In [105]:
reten

[23.029516704508595, 35.726083089823405, 42.4971363115693, 44.21211467988791]

In [122]:
# Summary table of retention
retention_summary = pd.DataFrame({
    "week": [32,33,34,35],
    "retention_rate": reten
})

retention_summary

week  retention_rate
0    32       23.029517
1    33       35.726083
2    34       42.497136
3    35       44.212115

### 9. What is the median time between user registration and first purchase? Give the answer in seconds (!) As an integer.

In [106]:
# getting timestamp for each user from user from initial point of registration
some_dict = dict()
for i, v in dfe.iterrows():
    if v["user_id"] not in some_dict.keys():
        some_dict[v["user_id"]] = [v["event_date"]]
    if v["user_id"] in some_dict.keys():
        if v["event_type"] == "purchase":
            some_dict[v["user_id"]].append(v["event_date"])

In [112]:
time_to_purchase = list(some_dict.values())

In [119]:
# calculating time difference 
time_diff = list()
for t in time_to_purchase:
    if len(t) >= 2:
        diff = t[0] - t[1]
        time_diff.append(diff)

In [121]:
# Median value
np.median(time_diff)

Timedelta('-6 days +23:13:46')

### Median time between registration and purchase is 6 days +23:13:46

### In seconds: 602, 026

518400
82800
780
46

##

## Task 3 - 30 Minutes
How would you answer the student's question below? Your task is to get your message across in such a way that a beginner can understand your explanation. You can do this any way you want (pictures, GIFs, metaphors, anything) so long as it makes your explanation clear.
Answer the question: "What is the difference between DataFrame and Series?"
Indicate how much time you spent completing this task.

### Answer
The python programming Language has a library called Pandas and Pandas is a powerful tool to handle various data formats. Pandas provides two data structures that it shapes or processes data into human readable form; these are the DataFrame and Series.
One major difference lies between a DataFrame and a Series.
A series is a one-dimensional data structure, whereas a dataframe is a two-dimensional data structure.

Take a vending machine for example; a series data structure is like a vending machine that can only give you a chocolate bar at a time as compared to a vending machine that can give you up to three or four chocolate bars at once. Cool right?
You can also see a series as a number of pairs having keys and values and a DataFrame can be similar to that of a spreadsheet. Below is a visual representation of series and DataFrames.

![alt text for screen readers](df_and_series.jpeg "Series and DataFrames")

You can also view a DataFrame as a combination of Series

## Task 4 - 1 hour

You are given two random variables X and Y.
E(X) = 0.5, Var(X) = 2
E(Y) = 7, Var(Y) = 3.5
cov (X, Y) = -0.8
Find the variance of the random variable Z = 2X - 3Y

### Solution

First, using the formula for variance of the linear combination of two variables x and y
given as 

$ Var(aX + bY) = a^2Var(X) + b^2Var(Y) + 2abCov(X,Y)$

Going by the above formula and substituting in the appropraite values:

$ Var(Z) = Var(2X - 3Y) = 2^2Var(X) + 3^2Var(Y) + 2 * 3 * 2Cov(X, Y) $

Substituting known values Var(X) = 2, Var(Y) = 3.5, cov(X,Y) = -0.8:

$ Var(Z) = (4 * 2) + (9 * 3.5) + 12(-0.8) $

$ Var(z) = 8 + 31.5 - 9.6 $

$ Var(z) = 29.9 $

Hence Variance $ Var(aX + bY) = Var(Z) = 29.9$


## Task 5 - 1 hour

Omer trained a linear regression model and tested its performance on a test sample of 500
objects. On 400 of those, the model returned a prediction higher than expected by 0.5, and on
the remaining 100, the model returned a prediction lower than expected by 0.7.
What is the MSE for his model?
Limor claims that the linear regression model wasn't trained correctly, and we can do improve
it by changing all the answers by a constant value. What will be her MSE?
You can assume that Limor found the smallest error under her constraints.
Return two values - Omer's and Limor's MSE.

### Solution
taking the error on the first 400 as 0.5 and on the remaining 100 as 0.7
we can assume from the formula given as:

$$ MSE = \dfrac{1}{n} \sum(y_t - y_p)^2$$

Total sample = 500
$$ MSE = \dfrac{1}{500} \sum 400(0.5)^2 + 100(0.7)^2$$

$$ MSE = \dfrac{1}{500} \sum (100 + 49)$$

$$ MSE = 0.298$$


The MSE assuming constant error across all values is 0.298

However, further information needs to be provided to further determine Limor's theory